added 'bootstrap' imputation strategy (renamed from 'random')

kootenpv opened the pull request.

kootenpv committed:
added 'random' imputation strategy

giorgiop commented on sklearn/preprocessing/imputation.py:
```diff
@@ -1,4 +1,4 @@
-# Authors: Nicolas Tresegnie <nicolas.tresegnie@gmail.com>
+# Authors: Nicolas Tresegnie <nicolas.tresegnie@gmail.com>, Pascal van Kooten
```
Could  you please move it to a new line?

giorgiop commented on sklearn/preprocessing/imputation.py:
```diff
@@ -248,6 +251,14 @@ def _sparse_fit(self, X, strategy, missing_values, axis):
 
                 return most_frequent
 
+            elif strategy == "random":
```
You should also update the comment above `# Median + Most frequent`, and add a comment here to be consistent.

giorgiop commented on sklearn/preprocessing/imputation.py:
```diff
@@ -299,6 +310,20 @@ def _dense_fit(self, X, strategy, missing_values, axis):
 
             return most_frequent
 
+        elif strategy == "random":
```
same as above.

kootenpv committed:
fixed 'Authors', added missing comments

giorgiop commented:
Not related with your PR: fit has `y` as an argument that is never used. Could you please either uniform the docstring to other transformer (example [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/preprocessing/data.py#L552))?

giorgiop commented:
More importantly: I think you should seed the random number generator. Please see [here](http://scikit-learn.org/stable/developers/utilities.html#validation-tools). Probably the best way is to write an additional class' attribute `random_state`.

kootenpv committed:
added 'y' as argument for compatability with Pipeline; communicated about it in docstring

kootenpv commented:
@giorgiop Thanks! I was looking at the examples, and noticed [a case](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/preprocessing/data.py#L347) where transform only accepts `X`. I recently rolled my own transformer and indeed also added the `y` argument since I couldn't see any real downsides. I added it in a new commit.

giorgiop commented:
Question: what's the rational for picking *one* random number to fill all the missing `m` ones in a column, instead of sampling `m` of them at random? I am just trying to understand your use case. I guess a random imputation strategy may be used as baseline.

GaelVaroquaux commented:
Indeed, I would be interested in publications or at least webpages
detailing the use that you have in mind.

kootenpv commented:
I tested it only on a single array, so maybe there is a mistake there? Are you only talking about `_sparse`?

The use case I have in mind is:

For each column, rather than taking the mean of the non-nan values and impute that for each missing value, instead, sample with replacement at once a value for each missing value.

glouppe commented:
I think this is a good idea, however I would suggest renaming this strategy to something more precise than "random". I think "bootstrap" would be more appropriate.

kootenpv commented:
I was considering "donor", but "bootstrap" might be even better. The sparse case seems to have a weird behavior, but I'm working on it.

I'll update the name to "bootstrap".

kootenpv commented:
I'm really confused what it means to impute in a sparse matrix. Is the following correct:

- Let's consider 0 to be the missing value for a sparse matrix (since it can only have 0 and 1).
- When imputing the average, it would calculate the proportion of missing for each column. If this is <50% it says the missing values will be on average 1, and if >50% it will say the value should be 0?

glouppe commented:
For sparse matrices, if `missing_values == 0`, then all zero entries (i.e., all the blanks in the sparse matrix) have to be imputed, leading to a dense matrix. If `missing_values != 0`, then all zero entries are considered to be actual zeros, which means that they should not be touched. Only some of the stored values (those correspond to `missing_values`) need to be imputed. See eg https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/preprocessing/imputation.py#L188

kootenpv commented:
@glouppe Yea, so either 0 or 1 gets imputed right (or whatever masked value for 0 or 1 might be used). It's binary, just a mirrored case.

Considering `missing_values==0`, what might be a resulting dense matrix?

glouppe commented:
Sparse matrices are not binary. They can stored any float/int number.

kootenpv commented:
@glouppe Now it makes sense, I'll let it sink in and work on a fix for sparse.

giorgiop commented:
I suggest you to start by adding the necessary tests in `test_imputation.py`. It should be easy to extend what we do to the new imputation option. What I meant before:
```
X = np.arange(9).reshape(3,3)
X
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

random = np.empty(X.shape[0])
for i, row in enumerate(X[:]):
    random[i] = np.random.RandomState(0).choice(row)
random
array([ 0.,  3.,  6.])
```
This only give you 1 random value per row (or column), but I think you intend to sample with replacement as many value as the ones to be imputed. Right? That loop is fine with the other imputation strategies, not with this one.

thomasjpfan commented:
Thank you for working on this. I am closing this PR because it was superseded by https://github.com/scikit-learn/scikit-learn/pull/11368.
